src/repository.c

```diff
@@ -2477,6 +2477,65 @@ int git_repository_state_cleanup(git_repository *repo)
 	return git_repository__cleanup_files(repo, state_files, ARRAY_SIZE(state_files));
 }
 
+int git_repository__shallow_roots(git_array_oid_t *out, git_repository *repo)
+{
+	git_buf path = GIT_BUF_INIT;
+	git_buf contents = GIT_BUF_INIT;
+	int error, updated;
+	size_t start, end;
+
+	assert(out);
+
+	if ((error = git_buf_joinpath(&path, repo->path_repository, "shallow")) < 0)
+		return error;
+
+	error = git_futils_readbuffer_updated(&contents, git_buf_cstr(&path), &repo->shallow_checksum, &updated);
+	git_buf_free(&path);
+
+	if (error < 0)
+		return error;
+
+	if (!updated) {
+		goto unchanged;
+	}
+
+	git_array_clear(repo->shallow_oids);
+
+	start = end = 0;
+	while (end < contents.size) {
+		if (contents.ptr[end] == '\n' || end == contents.size) {
+			if (end - start == GIT_OID_HEXSZ) {
+				git_oid *oid = git_array_alloc(repo->shallow_oids);
+
+				error = git_oid_fromstrn(oid, contents.ptr + start, GIT_OID_HEXSZ);
+				if (error < 0)
+					return error;
+			}
+			start = end + 1;
+		}
+		end++;
```

Comment:
The documentation about the shallow file format is pretty terse (what I know of it is what's in `Documentation/technical/shallow.txt`), and I wanted to know if it's acceptable to look at the source (since licensing).

I rebased and pushed a version that uses `git__strsep`, which really feels cleaner. I do have a point to make : our implementation is "non-conforming" (for some definition of it :wink:). Here's the relevant tidbit from my OS X man pages (which I was referring to when I wrote it) :

> The location of the next character after the delimiter character (or NULL, if the end of the string was reached) is stored in *stringp.  The original value of *stringp is returned.

Ours just leave `*stringp` as it is and returns `NULL` in that case, which can complicate parsing files where newlines-at-EOF are not guaranteed.

If that's expected, allow me to either 1) fix it or 2) add a comment on top of that function so you don't blindly trust your manpages (as I did).

For the record, the relevant changes are :

``` C
while ((line = git_strsep(&buffer, "\n")) != NULL) {
    /* stuff with line */
}
if (*buffer) {
    /* missing EOL */
}
```
to
``` C
while ((line = git_strsep(&buffer, "\n")) != NULL) {
    if (buffer == NULL)
        /*missing EOL */;
    /* stuff with line */
}
```